include/nbl/asset/ICPUGraphicsPipeline.h

```diff
@@ -13,91 +13,105 @@
-                return nullptr;
-            auto retval = new ICPUGraphicsPipeline(params);
-            for (const auto spec : params.shaders)
-            if (spec.shader)
-				retval->setSpecInfo(spec);
+        explicit ICPUGraphicsPipeline(const ICPUPipelineLayout* layout)
+            : base_t(layout, {}, {})
+            {}
+        
+        static core::smart_refctd_ptr<ICPUGraphicsPipeline> create(const ICPUPipelineLayout* layout)
+        {
+            auto retval = new ICPUGraphicsPipeline(layout);
             return core::smart_refctd_ptr<ICPUGraphicsPipeline>(retval,core::dont_grab);
-		}
-
-		constexpr static inline auto AssetType = ET_GRAPHICS_PIPELINE;
-		inline E_TYPE getAssetType() const override { return AssetType; }
-		
-		inline size_t getDependantCount() const override
-		{
-			auto stageCount = 2; // the layout and renderpass
-			for (const auto& stage : m_stages)
-			if (stage.shader)
-				stageCount++;
-			return stageCount;
-		}
-
-		// extras for this class
-		inline const SCachedCreationParams& getCachedCreationParams() const {return base_t::getCachedCreationParams();}
+        }
+
+        inline base_t* clone_impl(core::smart_refctd_ptr<const ICPUPipelineLayout>&& layout, uint32_t depth) const override final
+        {
+            auto* newPipeline = new ICPUGraphicsPipeline(layout.get());
+            for (auto i = 0; i < GRAPHICS_SHADER_STAGE_COUNT; i++)
+                newPipeline->m_specInfos[i] = m_specInfos[i];
+            newPipeline->m_params = m_params;
+            newPipeline->m_renderpass = m_renderpass;
+            
+            for (auto specInfo_i = 0u; specInfo_i < m_specInfos.size(); specInfo_i++)
+            {
+                newPipeline->m_specInfos[specInfo_i] = newPipeline->cloneSpecInfo(m_specInfos[specInfo_i], depth);
+            }
+
+            return newPipeline;
+        }
+
+        constexpr static inline auto AssetType = ET_GRAPHICS_PIPELINE;
+        inline E_TYPE getAssetType() const override { return AssetType; }
+        
+        inline size_t getDependantCount() const override
+        {
+            auto stageCount = 2; // the layout and renderpass
+            for (const auto& info : m_specInfos)
+            {
+              if (info.shader)
+                stageCount++;
+            }
+            return stageCount;
+        }
+
         inline SCachedCreationParams& getCachedCreationParams()
         {
             assert(isMutable());
             return m_params;
         }
 
+        inline virtual std::span<SShaderSpecInfo> getSpecInfo(hlsl::ShaderStage stage) override final
+        {
+            const auto stageIndex = stageToIndex(stage);
+            if (isMutable() && stageIndex != -1)
+            {
+                return { &m_specInfos[stageIndex], 1 };
+            }
+            return {};
+        }
+
+        inline virtual bool valid() const override final
+        {
+            if (!m_layout) return false;
```

Comment:
maybe `valid()` should be in `IAsset` and `ICPUPipeline` should already override and check the layout presence